invokeai/backend/generator/base.py

```diff
@@ -17,14 +21,259 @@
 from accelerate.utils import set_seed
 from diffusers import DiffusionPipeline
 from tqdm import trange
+from typing import List, Type, Iterator
+from dataclasses import dataclass, field
+from diffusers.schedulers import SchedulerMixin as Scheduler
 
-import invokeai.assets.web as web_assets
 from ..util.util import rand_perlin_2d
+from ..safety_checker import SafetyChecker
+from ..prompting.conditioning import get_uc_and_c_and_ec
+from ..model_management.model_manager import ModelManager
+from ..stable_diffusion.diffusers_pipeline import StableDiffusionGeneratorPipeline
 
 downsampling = 8
-CAUTION_IMG = "caution.png"
 
+@dataclass
+class InvokeAIGeneratorBasicParams:
+    seed: int=None
+    width: int=512
+    height: int=512
+    cfg_scale: int=7.5
+    steps: int=20
+    ddim_eta: float=0.0
+    model_name: str='stable-diffusion-1.5'
+    scheduler: int='ddim'
+    precision: str='float16'
+    perlin: float=0.0
+    threshold: int=0.0
+    h_symmetry_time_pct: float=None
+    v_symmetry_time_pct: float=None
+    variation_amount: float = 0.0
+    with_variations: list=field(default_factory=list)
+    safety_checker: SafetyChecker=None
+
+@dataclass
+class InvokeAIGeneratorOutput:
+    '''
+    InvokeAIGeneratorOutput is a dataclass that contains the outputs of a generation
+    operation, including the image, its seed, the model name used to generate the image
+    and the model hash, as well as all the generate() parameters that went into 
+    generating the image (in .params, also available as attributes)
+    '''
+    image: Image
+    seed: int
+    model_name: str
+    model_hash: str
+    params: dict
+
+    def __getattribute__(self,name):
+        try:
+            return object.__getattribute__(self, name)
+        except AttributeError:
+            params = object.__getattribute__(self, 'params')
+            if name in params:
+                return params[name]
+        raise AttributeError(f"'{self.__class__.__name__}' has no attribute '{name}'")
```

Comment:
You're quite right and I'll fix this. One issue is that I made `params` into a `dict` so that it could hold the union of the basic parameters fields (which are a dataclass object) and the various image generation-specific parameters, such as the img2img `strength`. So you'd have to access the fields like `output.params["strength"]` . I  think a good compromise would be to make the params a Namespace (or namespace-like) object so that `output.params.strength` works. Would that work?